src/tests/test_referral.cairo

```diff
@@ -72,287 +78,316 @@ fn V2_CLASS_HASH() -> starknet::class_hash::ClassHash {
-    let erc20 = deploy_erc20(recipient: OWNER(), initial_supply: u256 { low: 100000, high: 0 });
-    let naming = deploy_naming();
-    setup(naming.contract_address, erc20.contract_address, u256  { low: 1, high: 0 }, share:  u256 { low: 10, high: 0 });
+    let default_comm = 10;
+    let price_domain = 1000;
+    let (erc20, naming, referral) = setup(1, default_comm);
 
     // It should test upgrading implementation from a non-admin account
-    testing::set_caller_address(OTHER());
-    Referral::upgrade(V2_CLASS_HASH());
+    testing::set_contract_address(OTHER());
+    referral.upgrade(V2_CLASS_HASH());
 }
 
 #[test]
 #[available_gas(20000000)]
-#[should_panic(expected: ('Caller is the zero address',))]
+#[should_panic(expected: ('Caller is the zero address', 'ENTRYPOINT_FAILED'))]
 fn test_upgrade_fail_from_zero() {
-    let erc20 = deploy_erc20(recipient: OWNER(), initial_supply: u256 { low: 100000, high: 0 });
-    let naming = deploy_naming();
-    setup(naming.contract_address, erc20.contract_address, u256  { low: 1, high: 0 }, share:  u256 { low: 10, high: 0 });
+    let default_comm = 10;
+    let price_domain = 1000;
+    let (erc20, naming, referral) = setup(1, default_comm);
 
     // It should test upgrading implementation from the zero address
-    Referral::upgrade(V2_CLASS_HASH());
+    referral.upgrade(V2_CLASS_HASH());
+}
+
+#[test]
+#[available_gas(20000000)]
+fn test_add_rec_commission() {
+    let default_comm = 10;
+    let price_domain = 1000;
+
+    let (erc20, naming, referral) = setup(1, default_comm);
+
+    testing::set_contract_address(OWNER());
+
+    // It sends ETH to referral contract and then withdraw this amount from the contract
+    let initial_balance = 10000;
+    erc20.transfer_from(OWNER(), USER_A(), initial_balance);
+    erc20.transfer_from(OWNER(), USER_B(), initial_balance);
+    erc20.transfer_from(OWNER(), USER_C(), initial_balance);
+
+    // It should test calling add_commission from the naming contract & add the right commission
+    testing::set_contract_address(naming.contract_address);
+    assert(referral.get_balance(USER_A()) == 0, 'Init balance is incorrect');
+
+    // B referred by C
+    referral.add_commission(price_domain, USER_C(), USER_B());
+    let initial_expected = (price_domain * default_comm) / 100;
+    assert(referral.get_balance(USER_B()) == 0, 'Balance of B is incorrect');
+    assert(referral.get_balance(USER_C()) == initial_expected, 'Balance of C is incorrect');
+
+    // A referred by B
+    referral.add_commission(price_domain, USER_B(), USER_A());
+
+    assert(referral.get_balance(USER_A()) == 0, 'Balance of A is incorrect');
+    assert(referral.get_balance(USER_B()) == initial_expected, 'Balance of B is incorrect');
+    assert(
+        referral.get_balance(USER_C()) == initial_expected + initial_expected / 2,
+        'Balance of C is incorrect'
+    );
+}
+
+
+#[test]
+#[available_gas(20000000)]
+fn test_add_rec_circular_commission() {
```

Comment:
in this test, at this end, shouldn't you check that` referral.get_balance` gives the correct result for each user ? as the goal is to ensure they don't receive more than once

Reply:
Well it was mostly to unsure it doesn't loop forever as this was the risk, but I will improve the test